lib/client.js

```diff
@@ -1538,7 +1538,12 @@ function write (client, request) {
     contentLength = null
   }
 
-  if (request.contentLength !== null && request.contentLength !== contentLength) {
+  // https://github.com/nodejs/undici/issues/2046
+  // A user agent may send a Content-Length header with 0 value, this should be
+  // allowed.
+  const bypassDelete = method === 'DELETE' && contentLength === null && request.contentLength === 0
+
+  if (!bypassDelete && request.contentLength !== null && request.contentLength !== contentLength) {
```

Comment:
Two things:

1. We need to replicate this to `writeH2` otherwise will lead to inconsistencies
2. I have mixed feelings about doing this **only for DELETE** as does feel like a one-time-only fix; overall several methods allow an empty body or completely forbid it. I'm open to relaxing it for those payloads that do not expect a body or the body is optional. Should we consider a better strategy for these scenarios?

cc: @mcollina @ronag @KhafraDev

Reply:
Yea. I don't like this approach either.

Reply:
would it make more sense not to remove the content-length header if it's 0?

Reply:
```
   A user agent SHOULD NOT send a
   Content-Length header field when the request message does not contain
   a payload body and the method semantics do not anticipate such a
   body.
```

I think it's fine for DELETE only (or maybe re-evaluate which methods this should apply to, because the RFC doesn't forbid DELETE requests from having a body).

```
content received in a DELETE request has no generally defined semantics, cannot alter the meaning or target of the request, and might lead some implementations to reject the request and close the connection because of its potential as a request smuggling attack
```

Reply:
FWIW, here's the Go's logic for sending content-length
https://github.com/golang/go/blob/e9379a8f8bd5f00463f67fe7e052cbcb262b169a/src/net/http/transfer.go#L253

Reply:
>maybe re-evaluate which methods this should apply to, because the RFC doesn't forbid DELETE requests from having a body

I'm pro this approach; I believe is better to reevaluate and relax it for those scenarios where the `payload` is not expected or is optional (e.g. `DELETE` is optional, as well as `GET` and even `HEAD`).

The approach from `Go` doesn't seem bad, it can be a good starting point

Reply:
I made one more pass based on your suggestion, @metcoder95 @KhafraDev